uv/manager/manager.py

```diff
@@ -0,0 +1,106 @@
+from uv.reporter import AbstractReporter
+
+
+class MeasurementManager:
+  """Class for a MeasurementManager.  This manager is instantiated with a
+  particular reporter, which it will use to report whatever measurements is
+  conducts.
+
+  The MeasurementManager has a set of measurements---stored in a dictionary
+  (described below)---which are to be carried out at a certain frequency, which
+  is specific to each measurement.
+
+  Whenever the MeasurementManager's process() function is called, with a
+  step number provided as an argument, the MeasurementManager decides which
+  measurements are required at each step, carries them out, and reports them.
+
+  In order to carry out the measurement, the Manager has a state (Dict) at each
+  step, which contains all the information necessary to perform the measurement.
+  Part of this state is static (e.g., the training dataset), while part can
+  change at each step.  This dynamic state is passed along in the process() call.
+
+  Args:
+    static_state: Dict of variables which will be used to compute the measured
+                  values.  The static_state is fixed throughout time.
+    reporter: Reporter which will report the measurements.
+  """
+
+  def __init__(self, static_state: Dict[str, Any], reporter: AbstractReporter):
+    self.static_state = static_state
+    self.reporter = reporter
+    self.measurements = {}
+
+  def add_measurement(self, measurement_spec: Dict[str, Union[str, int,
+                                                              Callable]]):
+    """Adds a measurement specification to the MeasurementManager.
+    Args:
+      measurement_spec: Dict containing keys:
+        'name': String, the name of the measurement, e.g. 'training_loss'
+        'interval': Integer, how often to measure
+        'function': Function, taking in the state_dictionary, and returning the
+                              measured value
+    """
+    name = measurement_spec['name']
+    self.measurements[name] = {
+        'interval': measurement_spec['interval'],
+        'function': measurement_spec['function']
+    }
+
+  def process(self, step: int, dynamic_state: Dict[str, Any]):
+    """ At a given step, decide which measurements need to be
+    performed at this step and perform them
+
+    Args:
+      step: Integer, the number of the current step
+      dynamic_stat: Dict, the state variables necessary to perform the
+                    measurement
+    """
+
+    to_measure = self.required_measurements(step)
+
+    self.perform_specified_measurements(step, dynamic_state, to_measure)
+
+  def required_measurements(self, step: int) -> Iterable[str]:
+    """ Returns an iterable of measurement names which are to be
+    performed a the given step """
+
+    for name, spec in self.measurements.items():
+      if step % spec['interval'] == 0:
+        yield name
+
+  def perform_specified_measurements(self, step: int, dynamic_state: Dict[str,
```

Comment:
I would actually call this `process` or `measure` too. In that initial function, add 

```
measurement_list: Optional[Iterable[str]] = None
```

and then have

```python
if measurement_list is None:
  measurement_list = self.required_measurements(step)
```

Then proceed from there. you'll have one function and the interface will make more sense!

Reply:
Good idea, I've done that.